pypesto/sampling/geweke_test.py

```diff
@@ -0,0 +1,206 @@
+    k = np.floor((n - n_overlap) / (nw - n_overlap)).astype(int)
+    index = np.arange(nw)
+    # Normalizing scale factor
+    kmu = k * np.linalg.norm(w) ** 2
+    spectral_density = np.zeros((nfft))
+
+    for i in range(k):
+        xw = w * x[index]
+        index += (nw - n_overlap)
+        Xx = np.absolute(np.fft.fft(xw, n=nfft, axis=0)) ** 2
+        spectral_density += Xx
+
+    # Normalize
+    spectral_density = spectral_density * (1 / kmu)
+
+    n2 = np.floor(nfft / 2).astype(int)
+    spectral_density = spectral_density[0:n2]
+
+    return spectral_density
+
+
+def spectrum0(x: np.array):
+    """
+    Calculates the spectral density at frequency zero.
+
+    Parameters
+    ----------
+    x:
+        Fraction/fragment of the MCMC chain.
+
+    Returns
+    -------
+    spectral_density_zero:
+        Spectral density at zero.
+
+    """
+
+    n_samples, n_par = x.shape
+    spectral_density_zero = np.zeros((1, n_par))
+
+    for i in range(n_par):
+        spectral_density_zero[:, i] = \
+            spectrum(x[:, i], n_samples)[0]
+    return spectral_density_zero
+
+
+def calculate_zscore(chain: np.array,
+                     a: float = 0.1,
+                     b: float = 0.5):
+    """
+    Performs a Geweke test on a chain using the first
+    "a" fraction and the last "b" fraction of it for
+    comparison. Test for equality of the means of the
+    first a% and last b% of a Markov chain.
+
+    See:
+    Stephen P. Brooks and Gareth O. Roberts.
+    Assessing convergence of Markov chain Monte Carlo
+    algorithms. Statistics and Computing, 8:319--335, 1998.
+
+    Parameters
+    ----------
+    chain
+    a:
+        First fraction of the MCMC chain. Default 10%.
+    b:
+        Second fraction of the MCMC chain. Default 50%.
+
+    Returns
+    -------
+    z_score:
+        Z-score of the Geweke test.
+    p:
+        Significance level of the Geweke test.
+
+    """
+
+    nsimu, _ = chain.shape
+
+    # Define First fraction
+    index_a = np.floor(a * nsimu).astype(int)
+    # Define Second fraction
+    index_b = nsimu - np.floor(b * nsimu).astype(int) + 1
+
+    # Check if appropiate indexes
+    if (index_a + index_b) / nsimu > 1:
+        raise ValueError("Sample size too small to "
+                         "meaningfully extract subsets "
+                         "for Geweke's test.")
+
+    # Mean of First fraction
+    mean_a = np.mean(chain[0:index_a, :], axis=0)
+    # Mean of Second fraction
+    mean_b = np.mean(chain[index_b:, :], axis=0)
+
+    # Spectral estimates for variance
+    spectrum_a = spectrum0(chain[0:index_a, :])
+    spectrum_b = spectrum0(chain[index_b:, :])
+
+    # Calculate z-score
+    z_score = (mean_a - mean_b) / (np.sqrt(
+        spectrum_a / index_a + spectrum_b /
+        (nsimu - index_b + 1)
+    ))
+    # Calculate significance (p value)
+    p = 2 * (1 - norm.cdf(np.absolute(z_score)))
+
+    return z_score, p
+
+
+def burn_in_by_sequential_geweke(chain: np.array,
```

Comment:
could not find the reference implementation for this. please add, then I can check.

Reply:
The reference to this is an script from Benjamin Ballnuss. I cannot find a further reference to this. (I tried to attach it here but i am getting errors from git)

Reply:
ok, I have it.